detectors/gcp/cloud-function_test.go

```diff
@@ -0,0 +1,175 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gcp
```

Comment:
These tests skirt the issue where a user can create a bad instance of `CloudFunction` with `var CloudFunction{}` by reaching into the internals of the type. Can we change these tests to run external to `gcp` so they are evaluated from outside user's perspective. It doesn't seem there is anything being done in here that cannot be done from that perspective once the type initialization is cleaned up.

```suggestion
package gcp_test
```

Reply:
Correct, will start on test cases from outside user's perspective.
Thankyou @MrAlias

Reply:
> Now it can only be constructed via, contructor function.

Not if you test from within the package. Please make the suggested change to move tests out of the `gcp` package.

Reply:
I can do that, but, doing moving tests out of the gcp package will make it difficult to insert mock implementations, wherever necessary.
As, we have already made cloudFunction struct private, user's won't be able to create a bad instance of cloudFunction, as of now, they only have one option to create instance for cloudFunction, that is through using constructor function.

Reply:
@MrAlias , please let me know your thoughts on this.